### ISSUE TITLE
404 for OPTIONS requests when using `HandlerFunc` even when `OptionsPassthrough` is `false`

### ISSUE THREAD

**Issue by turtlemonvh** [rs/cors#21]
I'm not sure how general this is, but in the framework I'm using ([gin](https://github.com/gin-gonic/gin), which is very similar to [martini](https://github.com/rs/cors/blob/master/examples/martini/server.go)), when I use `c.HandlerFunc` I get 404s for OPTIONS requests, even with `OptionsPassthrough` set to `false`.

I was able to fix this in the short term by always setting the status code to 200 for valid OPTIONS requests.

Here's what I had before that failed with a 404

```
# Fails with 404 on OPTIONS request
r := gin.Default()
r.Use(gin.WrapF(c.HandlerFunc))
```

Here's what works for OPTIONS requests

```
makeCorsHandler := func(c *cors.Cors) func(http.ResponseWriter, *http.Request) {
    return func(w http.ResponseWriter, r *http.Request) {
        c.HandlerFunc(w, r)
        // Allow it to return to avoid a 404
        if r.Method == "OPTIONS" && w.Header().Get("Access-Control-Allow-Origin") == r.Header.Get("Origin") {
            w.WriteHeader(http.StatusOK)
        }
    }
}
```

Really I should be checking if `optionPassthrough` is `false`, but I don't have access to that property here.  Also, checking that `Access-Control-Allow-Origin` is set to the correct value is kind of a hack that just works because that header value [isn't set until the request has made it past all the relevant cross domain checks](https://github.com/rs/cors/blob/45d446e551b0020074e771dee17d2bb2fd2e9b44/cors.go#L279).

However you want to handle it, making the `OptionsPassthrough` option make more sense when using `HandlerFunc` would be helpful.  That may mean just documenting this case, or providing an option to handle this automatically by always returning 200s.  I couldn't think of a very clean way to do it, hence the bug report and no PR. :)

Please let me know if there is any testing I can do to help.  I certainly appreciate your work on this library!

**Comment by rs** [comment]
I'm not sure to fully understand what's going on by just reading your issue. I will try to find some time to study the behavior with Gin. I leave this issue open in the meantime.

**Comment by shrinathaithal** [comment]
@turtlemonvh is it okay if 200 is returned for "every" case?

**Comment by DSpeichert** [comment]
Have the same exact problem using Macaron - also using HandlerFunc

**Comment by EwanValentine** [comment]
I'm using this set-up... 

```
c := cors.New(cors.Options{
		AllowedOrigins:     []string{"*"},
		AllowCredentials:   true,
		OptionsPassthrough: true,
	})

	handler := c.Handler(api.router)

	// Run
	log.Fatal(http.ListenAndServe(port, handler))
```
With Gorilla mux router, getting 404 OPTIONS

**Comment by turtlemonvh** [comment]
@shrinathaithal Sorry I didn't see this before.  I think that returning a 200 is always ok for OPTIONS requests if `OptionsPassthrough` is set to `false`, since this sort of implies that this library is handling OPTIONS requests for you.

It looks like that was fixed per #26, which solves my short term problems, at least.
